File: dashpilot-ios/dashpilot/DataSource/DemoDataSource.swift
import Foundation

/// Generates believable `CarState` values without any hardware or network
/// connection, so the dashboards can be demoed anywhere (Settings > Start
/// Demo Mode). Field encodings (gear, traffic light color, etc.) mirror the
/// ones the dash-apps already expect — see `gearMap`/`TL_COLOR_MAP` in
/// `dash-apps/web-vanilla/index.html` and the demo generator in
/// `dash-apps/web-expo/hooks/useCarState.ts`.
final class DemoDataSource: IDataSource {

    private var continuation: AsyncStream<CarState>.Continuation?
    private var task: Task<Void, Never>?

    let incomingMessages: AsyncStream<CarState>

    init() {
        var capturedContinuation: AsyncStream<CarState>.Continuation?
        incomingMessages = AsyncStream(bufferingPolicy: .bufferingNewest(1)) { continuation in
            capturedContinuation = continuation
        }
        self.continuation = capturedContinuation
    }

    /// `address` is unused — demo mode has nothing to connect to.
    func connect(address: String) {
        task = Task { [weak self] in
            var elapsed: TimeInterval = 0
            let dt: TimeInterval = 0.1
            while !Task.isCancelled {
                self?.continuation?.yield(Self.frame(at: elapsed))
                elapsed += dt
                try? await Task.sleep(for: .seconds(dt))
            }
        }
    }

    func disconnect() {
        task?.cancel()
        task = nil
        continuation?.finish()
    }

    /// One lap through cruise -> a gentle turn -> more cruise -> a red light
    /// that's actually approached and stopped at -> green -> back to cruise.
    private static let cycleLength: TimeInterval = 90
    private static let turnStart: TimeInterval = 25, turnEnd: TimeInterval = 33
    private static let cruise2End: TimeInterval = 50
    private static let approachEnd: TimeInterval = 68
    private static let stoppedEnd: TimeInterval = 75
    private static let goEnd: TimeInterval = 78
    // cycleLength is the end of the accelerate-back-to-cruise phase.

    /// Slowly drifting base speed, continuous across cycle boundaries so
    /// cruise phases never jump.
    private static func cruiseSpeed(at t: TimeInterval) -> Float {
        Float(70 + 20 * sin(t * 0.02))
    }

    private static func frame(at t: TimeInterval) -> CarState {
        var s = CarState()
        let cycleIndex = Int(t / cycleLength)
        let phase = t.truncatingRemainder(dividingBy: cycleLength)
        let turningLeft = cycleIndex % 2 == 0

        s.gear = 4 // Drive
        s.adasOn = true
        s.madsActive = true
        s.selfdriveActive = true
        s.experimentalMode = false
        s.changingLane = false
        s.anyDoorOpen = 0
        s.buckleStatus = 1

        // Speed limit sign only changes once per lap (rare), and only to a
        // realistic posted limit.
        let limits: [Float] = [50, 90, 130]
        s.fusedSpeedLimit = limits[cycleIndex % limits.count]

        switch phase {
        case ..<turnStart:
            // Cruise.
            s.egoSpeed = cruiseSpeed(at: t)
            s.egoSteeringAngle = Float(sin(t * 0.3) * 3) // gentle lane-keeping wiggle
            s.trafficLightColor = 0
            s.stopLineDist = 0

        case ..<turnEnd:
            // Smooth, gradual turn: steering eases out and back in, speed
            // dips a little through the corner, turn signal blinks toward it.
            let progress = (phase - turnStart) / (turnEnd - turnStart)
            let ease = sin(progress * .pi) // 0 -> 1 -> 0
            let direction: Float = turningLeft ? -1 : 1
            s.egoSteeringAngle = direction * Float(ease) * 28
            s.egoSpeed = cruiseSpeed(at: t) * Float(1 - 0.5 * ease)
            let blink: Float = phase.truncatingRemainder(dividingBy: 1.0) < 0.5 ? 1 : 0
            s.leftBlinker = turningLeft ? blink : 0
            s.rightBlinker = turningLeft ? 0 : blink
            s.trafficLightColor = 0
            s.stopLineDist = 0

        case ..<cruise2End:
            s.egoSpeed = cruiseSpeed(at: t)
            s.egoSteeringAngle = Float(sin(t * 0.3) * 3)
            s.trafficLightColor = 0
            s.stopLineDist = 0

        case ..<approachEnd:
            // Approaching a red light: decelerate smoothly to a full stop
            // exactly as the stop line reaches the car.
            let progress = (phase - cruise2End) / (approachEnd - cruise2End)
            let phaseStartT = t - (phase - cruise2End) // wall-clock time this phase began
            let startSpeed = cruiseSpeed(at: phaseStartT)
            s.egoSpeed = startSpeed * Float(1 - progress)
            s.egoSteeringAngle = 0
            s.trafficLightColor = 1 // Red
            s.stopLineDist = Float(45 * (1 - progress))

        case ..<stoppedEnd:
            // Stopped at the line, waiting for green.
            s.egoSpeed = 0
            s.egoSteeringAngle = 0
            s.trafficLightColor = 1 // Red
            s.stopLineDist = 0

        case ..<goEnd:
            // Light just turned green; still moving off.
            let progress = (phase - stoppedEnd) / (goEnd - stoppedEnd)
            s.egoSpeed = cruiseSpeed(at: t) * Float(progress) * 0.3
            s.egoSteeringAngle = 0
            s.trafficLightColor = 2 // Green
            s.stopLineDist = 0

        default:
            // Accelerating back up to cruise speed.
            let progress = (phase - goEnd) / (cycleLength - goEnd)
            s.egoSpeed = cruiseSpeed(at: t) * Float(0.3 + 0.7 * progress)
            s.egoSteeringAngle = 0
            s.trafficLightColor = 0
            s.stopLineDist = 0
        }

        s.accSetSpeed = cruiseSpeed(at: t) + 5

        s.leftBlindSpot = t.truncatingRemainder(dividingBy: 23) < 3 ? 1 : 0
        s.rightBlindSpot = t.truncatingRemainder(dividingBy: 29) < 3 ? 1 : 0
        s.laneDepartureWarning = 0
        s.sideCollisionWarning = 0

        // BMS / power — Tesla-ish pack, slowly discharging.
        s.fullPackEnergy = 75
        s.nominalEnergyRemaining = Float(max(10, 60 - t * 0.01))
        s.energyBuffer = 1.5
        s.maxRegenPower = 60
        s.maxDischargePower = 300
        s.packVoltage = Float(380 + sin(t * 0.2) * 5)
        s.packCurrent = s.egoSpeed * 0.4
        s.packTMin = 22
        s.packTMax = 28
        s.acTemp = 22
        s.odometer = Float(42000 + t * 0.01)

        return s
    }
}

File: dashpilot-android/app/src/main/java/com/softwiredtech/dashpilot/datasource/DemoDataSource.kt
package com.softwiredtech.dashpilot.datasource

import com.softwiredtech.dashpilot.datamodel.dash.CarState
import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.SupervisorJob
import kotlinx.coroutines.cancel
import kotlinx.coroutines.channels.BufferOverflow
import kotlinx.coroutines.delay
import kotlinx.coroutines.flow.Flow
import kotlinx.coroutines.flow.MutableSharedFlow
import kotlinx.coroutines.isActive
import kotlinx.coroutines.launch
import kotlin.math.cos
import kotlin.math.sin

/**
 * Scripted 90 second drive that loops forever: pull away, cruise at 50, accelerate onto a
 * 120 road with ADAS engaged, brake for a red light with regen, park and open a door.
 */
class DemoDataSource : IDataSource {

    private val _incoming = MutableSharedFlow<CarState>(
        replay = 1,
        extraBufferCapacity = 1,
        onBufferOverflow = BufferOverflow.DROP_OLDEST
    )
    override val incomingMessages: Flow<CarState> = _incoming

    private var scope: CoroutineScope? = null

    override fun connect(address: String) {
        disconnect()
        val s = CoroutineScope(SupervisorJob() + Dispatchers.Default)
        scope = s
        s.launch {
            val start = System.currentTimeMillis()
            var odometerKm = 48213f
            var lastT = 0f
            while (isActive) {
                val t = ((System.currentTimeMillis() - start) / 1000f) % LOOP_SECONDS
                val speed = speedAt(t)
                val accel = (speed - speedAt(t - 0.5f)) / 0.5f
                val dt = if (t >= lastT) t - lastT else 0f
                lastT = t
                odometerKm += speed / 3600f * dt

                val powerKw = (speed * 0.12f + accel * 4.5f).coerceIn(-60f, 250f)
                val adas = t in 30f..70f
                val red = t in 70f..80f

                _incoming.tryEmit(
                    CarState(
                        egoSteeringAngle = 12f * sin(t * 0.7f),
                        egoSpeed = speed,
                        leftBlinker = blink(t, 12f, 17f),
                        rightBlinker = blink(t, 45f, 50f),
                        gear = if (t < 2f || t >= 85f) 1f else 4f,
                        adasOn = adas,
                        leftBlindSpot = window(t, 14f, 19f),
                        rightBlindSpot = window(t, 47f, 52f),
                        fusedSpeedLimit = if (t < 25f || t >= 65f) 50f else 120f,
                        stopLineDist = if (red) ((80f - t) * 6f).coerceAtLeast(0f) else 0f,
                        trafficLightColor = when {
                            red -> 1f
                            t in 80f..83f -> 2f
                            else -> 0f
                        },
                        laneDepartureWarning = window(t, 58f, 60f),
                        sideCollisionWarning = 0f,
                        anyDoorOpen = window(t, 86f, 89f),
                        buckleStatus = if (t < 2f) 0f else 1f,
                        accSetSpeed = if (adas) 120f else 0f,
                        fullPackEnergy = 78f,
                        nominalEnergyRemaining = 52f - t * 0.02f,
                        energyBuffer = 3f,
                        maxRegenPower = 60f,
                        maxDischargePower = 250f,
                        packVoltage = 380f,
                        packCurrent = powerKw * 1000f / 380f,
                        packTMin = 22f + 0.5f * cos(t * 0.1f),
                        packTMax = 27f + 0.5f * cos(t * 0.1f),
                        odometer = odometerKm,
                        acTemp = 21f,
                        acTempRight = 21.5f,
                        hvacFanLevel = 3f,
                        hvacPowerState = 1f,
                        hvacAcMode = 1f,
                        vin = "5YJ3DEMO000000000"
                    )
                )
                delay(TICK_MS)
            }
        }
    }

    override fun disconnect() {
        scope?.cancel()
        scope = null
    }

    private fun speedAt(t: Float): Float = when {
        t < 0f -> 0f
        t < 10f -> ease(t / 10f) * 50f
        t < 25f -> 50f
        t < 40f -> 50f + ease((t - 25f) / 15f) * 70f
        t < 60f -> 120f
        t < 80f -> 120f * (1f - ease((t - 60f) / 20f))
        else -> 0f
    }

    private fun ease(x: Float): Float {
        val c = x.coerceIn(0f, 1f)
        return c * c * (3f - 2f * c)
    }

    private fun window(t: Float, from: Float, to: Float) = if (t in from..to) 1f else 0f

    private fun blink(t: Float, from: Float, to: Float) =
        if (t in from..to && ((t * 2.5f).toInt() % 2 == 0)) 1f else 0f

    private companion object {
        const val LOOP_SECONDS = 90f
        const val TICK_MS = 40L
    }
}
